src/crimson/osd/osd.cc

```diff
@@ -285,6 +287,40 @@ seastar::future<> OSD::_send_boot()
   return monc->send_message(m);
 }
 
+seastar::future<> OSD::_add_me_to_crush()
+{
+  if (!local_conf().get_val<bool>("osd_crush_update_on_start")) {
+    return seastar::now();
+  }
+  const double weight = [this] {
+    if (auto w = local_conf().get_val<double>("osd_crush_initial_weight");
+	w >= 0) {
+      return w;
+    } else {
+      auto total = store->stat().total;
+      return std::max(.00001, double(total) / double(1ull << 40)); // TB
+    }
+  }();
+  const CrushLocation loc{make_unique<CephContext>().get()};
+  logger().info("{} crush location is {}", __func__, loc);
+  string cmd = fmt::format(R"({{
+    "prefix": "osd crush create-or-move",
+    "id": {},
+    "weight": {:.4f},
+    "args": [{}]
+  }})", whoami, weight, loc);
+  return monc->run_command({cmd}, {}).then(
+    [this](int32_t code, string message, bufferlist) {
+      if (code) {
+	logger().warn("fail to add to crush: {} ({})", message, code);
+	throw std::runtime_error("fail to add to crush");
```

Comment:
i understand this is not *runtime_error*. will try to find/create a better exception type for this purpose.